File: src/main/java/br/com/pnp/compiler/semantic/SymbolTable.java
package br.com.pnp.compiler.semantic;

import br.com.pnp.compiler.model.construct.Construct;
import br.com.pnp.compiler.model.construct.Variable;

import java.util.*;

public class SymbolTable {
    private SymbolTable parent = null;
    private HashMap<String, Construct> symbols = new LinkedHashMap<>();

    SymbolTable() { }

    private SymbolTable(SymbolTable parent) {
        this.parent = parent;
    }

    public SymbolTable startNewScope() {
        return new SymbolTable(this);
    }

    public SymbolTable lastScope() {
        return parent;
    }

    public Construct tryGetValue(String key) {
        if (symbols.containsKey(key)) {
            return symbols.get(key);
        }
        if (parent != null) {
            return parent.tryGetValue(key);
        }
        return null;
    }

    public void tryPutValue(String key, Construct value) {
        if (!symbols.containsKey(key)) {
            symbols.put(key, value);
        }
    }

    public boolean existsInThisScope(String key) {
        return symbols.containsKey(key);
    }

    public boolean exists(String key) {
        if (!symbols.containsKey(key)) {
            return parent.exists(key);
        }
        return true;
    }

    public List<Variable> getVariables() {
        List<Variable> variables = new ArrayList<>();

        for (Construct entry : symbols.values()) {
            if (entry instanceof Variable) {
                variables.add((Variable)entry);
            }
        }

        return variables;
    }

    public Collection<Construct> getValues() {
        return symbols.values();
    }
}

File: src/main/kotlin/br/com/pnp/semantic/SymbolTable.kt
package br.com.pnp.semantic

import br.com.pnp.model.construct.Construct
import br.com.pnp.model.construct.Variable

class SymbolTable() {
    private var parent: SymbolTable? = null
    val symbols = LinkedHashMap<String, Construct>()

    val variables: List<Variable>
        get() = symbols.values.filterIsInstance<Variable>()
    val values: List<Construct>
        get() = symbols.values.toList()

    private constructor(parent: SymbolTable) : this() {
        this.parent = parent
    }

    fun startNewScope(): SymbolTable {
        return SymbolTable(this)
    }

    fun lastScope(): SymbolTable? {
        return parent
    }

    fun tryGetValue(key: String): Construct? {
        if (symbols.containsKey(key)) {
            return symbols[key]
        }
        return parent?.tryGetValue(key)
    }

    fun tryPutValue(key: String, value: Construct): Boolean {
        if (!symbols.containsKey(key)) {
            symbols[key] = value
            return true
        }
        return false
    }

    fun existsInThisScope(key: String): Boolean {
        return symbols.containsKey(key)
    }

    fun exists(key: String): Boolean {
        if (!symbols.containsKey(key)) {
            return parent?.exists(key) ?: false
        }
        return true
    }

    fun resetTable() {
        symbols.clear()
    }
}
